casper/src/main/resources/PoS.rhox

```diff
@@ -63,626 +65,682 @@ in {
-                              *removeQuarantinedCh)
-                          } |
-
-                          for(@newValidators <- newValidatorsCh;
-                              @(payments, newBonds, newWithdrawers, newRewards) <- removeQuarantinedCh;
-                              _ <- accDepositsDoneCh) {
-
-                            payValidators!(payments, *paymentDoneCh) |
-
-                            for (_ <- paymentDoneCh) {
-
-                              mvarResultCh!({
-                                "allBonds"        : newBonds,
-                                "activeValidators": newValidators,
-                                "committedRewards": newRewards,
-                                "withdrawers"     : newWithdrawers,
-                                "randomImages"    : {},
-                                "randomNumbers"   : {},
-                              }, (true, Nil))
-                            }
                           }
                         }
-                      } |
-
-                      new moveFunds in {
-                        contract moveFunds(@(unf, _, vault), returnCh) = {
-                          new balanceCh, unfAuthKeyCh, transferDoneCh in {
-                            @vault!("balance", *balanceCh) |
-                            for (@balance <- balanceCh) {
-                              if (balance > 0) {
-                                @RevVault!("unforgeableAuthKey", unf, *unfAuthKeyCh) |
-                                for (@authKey <- unfAuthKeyCh) {
-                                  @vault!("transfer", posRevAddress, balance, authKey, *transferDoneCh) |
-                                  for (@(true, Nil) <- transferDoneCh) {
-                                    returnCh!(true)
+                      } else {
+                        returnCh!((false, "Invalid system auth token"))
+                      }
+                    }
+                  }
+                } |
+                // Private method which signals the end of block processing.
+                contract PoS(@"closeBlock", @sysAuthToken, ackCh) = {
+                  new isValidTokenCh in {
+                    sysAuthTokenOps!("check", sysAuthToken, *isValidTokenCh) |
+                    for (@isValid <- isValidTokenCh) {
+                      if (isValid) {
+                        new changeEpoch, payWithdrawers,
+                            removeQuarantinedWithdrawers
+                        in {
+                          new blockDataCh, stateProcessCh,
+                              getBlockData(`rho:block:data`),
+                              accDepositsDoneCh, commitRewardsCh,
+                              newValidatorsCh, removeQuarantinedCh,
+                              paymentDoneCh
+                          in {
+                            getBlockData!(*blockDataCh) |
+                            runMVar!(*stateCh, *stateProcessCh, *ackCh) |
```

Comment:
This `runMVar` will _change_ the state so each block is conflicting. We should _peek_ state here and only change state on epoch change.
@Isaac-DeFrain I think you already done this when we were resolving mergeability issues. Can you include it in this PR?

Reply:
I must've missed that while merging conflicts... I'll make that change. Thanks for catching it!

Reply:
Currently, the epochLength is set to 1. So we're changing epochs at every closeBlock.

Reply:
Are you referring to test configuration for epochLength? Default in [config](https://github.com/rchain/rchain/blob/d8a8e7f/node/src/main/resources/reference.conf#L160) is 10000.

Reply:
Okay, that's great! I was referring to https://github.com/rchain/rchain/blob/dev/casper/src/test/scala/coop/rchain/casper/util/GenesisBuilder.scala#L54